contracts/proxy-deployer/src/contract_interactions.rs

```diff
@@ -0,0 +1,206 @@
+multiversx_sc::imports!();
+
+use multiversx_sc_modules::pause;
+
+use crate::config::{self, OngoingUpgradeOperation};
+
+#[multiversx_sc::module]
+pub trait ContractInteractionsModule: config::ConfigModule + pause::PauseModule {
+    #[endpoint(contractDeploy)]
+    fn contract_deploy(
+        &self,
+        template_address: ManagedAddress,
+        args: MultiValueEncoded<ManagedBuffer>,
+    ) -> ManagedAddress {
+        self.can_call_endpoint(None);
+        require!(
+            self.blockchain().is_smart_contract(&template_address),
+            "Template address is not a SC"
+        );
+
+        let (new_contract_address, _) = self.send_raw().deploy_from_source_contract(
+            self.blockchain().get_gas_left(),
+            &BigUint::zero(),
+            &template_address,
+            CodeMetadata::DEFAULT,
+            &args.to_arg_buffer(),
+        );
+
+        let caller = self.blockchain().get_caller();
+        let mut deployed_addresses = match self
+            .deployer_template_addresses(&caller)
+            .get(&template_address)
+        {
+            Some(addresses) => addresses,
+            None => ManagedVec::new(),
+        };
+        deployed_addresses.push(new_contract_address.clone());
+
+        self.deployer_contracts(&caller).add(&new_contract_address);
+        self.deployed_contracts_list_by_template(&template_address)
+            .update(|deployed_contracts| {
+                deployed_contracts.push(new_contract_address.clone());
+            });
+        self.deployer_template_addresses(&caller)
+            .insert(template_address, deployed_addresses);
+        self.deployers_list().insert(caller);
+
+        new_contract_address
+    }
+
+    #[endpoint(contractUpgrade)]
+    fn contract_upgrade(
+        &self,
+        contract_address: ManagedAddress,
+        template_address: ManagedAddress,
+        args: MultiValueEncoded<ManagedBuffer>,
+    ) {
+        self.can_call_endpoint(Some(contract_address.clone()));
+        require!(
+            self.blockchain().is_smart_contract(&contract_address),
+            "Contract address is not a SC"
+        );
+        require!(
+            self.blockchain().is_smart_contract(&template_address),
+            "Template address is not a SC"
+        );
+
+        self.send_raw().upgrade_from_source_contract(
+            &contract_address,
+            self.blockchain().get_gas_left(),
+            &BigUint::zero(),
+            &template_address,
+            CodeMetadata::DEFAULT,
```

Comment:
Is there a way where you can keep the previous CodeMetadata? Maybe after upgrade, the owner doesn't want to change them to default.

Reply:
Yes, we plan to have it working this way (keeping the previous CodeMetadata). But we need to wait for the upgrade + a new function in the framework.